crypto/rsa/rsa_oaep.c

```diff
@@ -234,25 +234,21 @@ int RSA_padding_check_PKCS1_OAEP_mgf1(unsigned char *to, int tlen,
     good &= constant_time_ge(tlen, mlen);
 
     /*
-     * Even though we can't fake result's length, we can pretend copying
-     * |tlen| bytes where |mlen| bytes would be real. Last |tlen| of |dblen|
-     * bytes are viewed as circular buffer with start at |tlen|-|mlen'|,
-     * where |mlen'| is "saturated" |mlen| value. Deducing information
-     * about failure or |mlen| would take attacker's ability to observe
-     * memory access pattern with byte granularity *as it occurs*. It
-     * should be noted that failure is indistinguishable from normal
-     * operation if |tlen| is fixed by protocol.
+     * Move the result in-place by |dblen|-|mdlen|-1-|mlen| bytes to the left.
+     * Then if |good| move |mlen| bytes from |db|+|mdlen|+1 to |to|.
+     * Otherwise leave |to| unchanged.
+     * Do that all in constant time.
      */
     tlen = constant_time_select_int(constant_time_lt(dblen - mdlen - 1, tlen),
                                     dblen - mdlen - 1, tlen);
-    msg_index = constant_time_select_int(good, msg_index, dblen - tlen);
-    mlen = dblen - msg_index;
-    for (mask = good, i = 0; i < tlen; i++) {
-        unsigned int equals = constant_time_eq(msg_index, dblen);
-
-        msg_index -= tlen & equals;  /* rewind at EOF */
-        mask &= ~equals;  /* mask = 0 at EOF */
-        to[i] = constant_time_select_8(mask, db[msg_index++], to[i]);
+    for (msg_index = 1; msg_index < dblen - mdlen - 1; msg_index <<= 1) {
```

Comment:
Is there a more detailed description about how this pattern works?
It is not obvious it does what it does.  It appears to be copying over the destination multiple times which is odd.

If not, perhaps a comment describing it?

Reply:
I figured it out the second time around.
It still wasn't obvious.

Reply:
Okay, will try to come up with a comment explaining the idea.

Reply:
A clever solution to a difficult problem :)

Reply:
I admit I have stolen the idea from @davidben

Reply:
How about something along this line:

"Copy the memory back in a way that does reveal the size of the data being copied via a timing side channel.  The requires copying parts of the buffer multiple times based on the bits set in the real length.  Clear bits do a non-copy with similar access pattern."

Reply:
opps, I did not see your comment above.  I guess I take yours and add the statement about complexity.

Reply:
I guess I can say "Clear bits do a non-copy with identical access pattern".

Reply:
You statement is good enough I think.  It makes it clear that it is intentional and gives enough information to work out what it is doing.